nomad/leader.go

```diff
@@ -948,6 +959,88 @@ func (s *Server) restorePeriodicDispatcher() error {
 	return nil
 }
 
+// restoreLockTTLTimers iterates the stored variables and creates a lock TTL
+// timer for each variable lock. This is used during leadership establishment
+// to populate the in-memory timer.
+func (s *Server) restoreLockTTLTimers() error {
+
+	varIterator, err := s.fsm.State().Variables(nil)
+	if err != nil {
+		return fmt.Errorf("failed to list variables for lock TTL restore: %v", err)
+	}
+
+	// Iterate the variables, identifying each one that is associated to a lock
+	// and adding a TTL timer for each.
+	for varInterface := varIterator.Next(); varInterface != nil; varInterface = varIterator.Next() {
+		if realVar, ok := varInterface.(*structs.VariableEncrypted); ok && realVar.IsLock() {
+
+			// The variable will be modified in order to show that it no longer
+			// is held. We therefore need to ensure we perform modifications on
+			// a copy.
+			realVarCopy := realVar.Copy()
+
+			s.createVariableLockTTLTimer(&realVarCopy)
+		}
+	}
+
+	return nil
+}
+
+// createVariableLockTTLTimer creates a TTL timer for the given lock. The
+// passed ID is expected to be generated via the variable NamespacedID
+// function.
+func (s *Server) createVariableLockTTLTimer(variable *structs.VariableEncrypted) {
+
+	// Adjust the given TTL by multiplier of 2. This is done to give a client a
+	// grace period and to compensate for network and processing delays. The
+	// contract is that a variable lock is not expired before the TTL expires,
+	// but there is no explicit promise about the upper bound so this is
+	// allowable.
+	lockTTL := variable.Lock.TTL * 2
```

Comment:
Multiplier of 2 is somewhat arbitrary, no? For very small TTLs this won't really give a good grace period.

Reply:
Why not use a factor?

Reply:
if the leader transition can take up to 2 ttls, will all the calls to renew will fail in that time?

Reply:
TTLs cannot be lower than a 10s time duration, and although arbitrary, I feel it's a good enough number to pick as does Consul. I am open to other suggestions though.

Reply:
I think if minimum is 10s then you're totally fine.